### PR TITLE
Bail early if remote wouldn't be retried

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/util/retryutils.py [hunk 1]
```diff
@@ -35,6 +35,29 @@ def __init__(self, retry_last_ts, retry_interval, destination):
...
+    if retry_timings:
+        retry_last_ts, retry_interval = (
+            retry_timings["retry_last_ts"], retry_timings["retry_interval"]
+        )
+
+        now = int(clock.time_msec())
+
+        if retry_last_ts + retry_interval > now:
+            defer.returnValue(True)
+    defer.returnValue(True)
```
Comment (richvdh) [inline synapse/util/retryutils.py]: This appears to be missing a False case.

File: synapse/util/retryutils.py [hunk 2]
```diff
@@ -35,6 +35,29 @@ def __init__(self, retry_last_ts, retry_interval, destination):
 
 
 @defer.inlineCallbacks
+def would_retry(destination, clock, store):
+    """Check if ``get_retry_limiter`` would throw. This is purely to allow
+    things to bail early if the resulting request would not otherwise be sent
+    """
```
Comment (richvdh) [inline synapse/util/retryutils.py]: please can we document types of args and return values, even if it's "obvious". I'm fed up with figuring it all out from first principles in the synapse codebase. You can even c&p most of it from `get_retry_limiter` in this case.

File: synapse/federation/transaction_queue.py [hunk 3]
```diff
@@ -303,9 +303,13 @@ def _attempt_new_transaction(self, destination):
             )
             return
 
+        pending_pdus = []
         try:
             self.pending_transactions[destination] = 1
 
+            if not (yield would_retry(destination, self.clock, self.store)):
```
Comment (richvdh) [inline synapse/federation/transaction_queue.py]: given we logger.debug when we get a `NotRetryingDestination` exception (a block of code which is now almost redundant, btw), should we do so here? should we in fact throw a NotRetryingDestination here?

File: synapse/util/retryutils.py [hunk 4]
```diff
@@ -35,6 +35,29 @@ def __init__(self, retry_last_ts, retry_interval, destination):
 
 
 @defer.inlineCallbacks
+def would_retry(destination, clock, store):
```
Comment (richvdh) [inline synapse/util/retryutils.py]: Of course another way to do this is just to call `get_retry_limiter`, and throw away the result, and see if you get an exception.

**Comment by erikjohnston** [comment]
You raise excellent points.

File: synapse/federation/transaction_queue.py [hunk 5]
```diff
@@ -303,9 +303,15 @@ def _attempt_new_transaction(self, destination):
             )
             return
 
+        pending_pdus = []
         try:
             self.pending_transactions[destination] = 1
 
+            # This will throw if we woldn't retry. We do this here so we fail
```
Comment (richvdh) [inline synapse/federation/transaction_queue.py]: woldn't

**Pull request merged**